presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -598,17 +612,50 @@ else if (redistributeWrites) {
                     !source.getProperties().isCompatibleTablePartitioningWith(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                     !(source.getProperties().isRefinedPartitioningOver(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                             canPushdownPartialMerge(source.getNode(), partialMergePushdownStrategy))) {
```

Comment:
why does `canPushdownPartialMerge(source.getNode(), partialMergePushdownStrategy)` matter? Is shuffle more expensive compared with low parallelism of write?

Reply:
Let's say you want to write to a bucketed table with `1024` buckets when joining 2 tables that are bucketed into `4096` buckets. We don't want to process the join with only `1024` tasks, as we will be at risk of running out of memory. That's why we have this check in place.